service/storage_service.cc

```diff
@@ -1181,6 +1192,7 @@ std::vector<canonical_mutation> storage_service::build_mutation_from_join_params
     node_builder.set("request_id", params.request_id);
     topology_request_tracking_mutation_builder rtbuilder(params.request_id);
     rtbuilder.set("initiating_host",_group0->group0_server().id().uuid())
+             .set("request_type", params.replaced_id ? topology_request::replace : topology_request::join)
```

Comment:
Yeah, this can only be done if all nodes have the new column, so we need a cluster feature.

Reply:
But can't this information be already inferred from existing stuff we already store?

We have the `topology_request` already inside system.topology, and a reference to the request. (request ID)

It looks reduntant to store it again inside the requests table. Unless I'm missing something?

Reply:
But it only keeps info about pending requests, no?

Reply:
I'm not sure. @gleb-cloudius ?

Reply:
Request ID in topology exist only while request is running. But I did not add request type to the table because group0 history table can be used to figure out what the operation was (group0 state id is used as a key for both). Why do we need this field in the table?

Reply:
group 0 history is transient though, and the description column is not designed to be used by algorithm but instead read by a human (it's an english human-readable string). I know that some tests are (ab)using the history but there is no promise for backward compatibility of the descriptions etc.

Reply:
We want to be able to present the result of a topology request to the user, via the task-manager API. It is not a problem if request table is TTLed, we don't need information on completed requests going back to the beginning of times. We just want to make the finish status of a request available to query, after a reasonable amount of time. Think of an admin who kicked off a topology request, then went for a "quick" coffe, 2h later when he comes back to his machine, he wants to check what the result of that request was.

Reply:
Yes, but then the info will be in the history table still. Anyway it is not a big deal to add the field to the table, but as you see it requires features and stuff by this point, so if it is possible to avoid it it will simplify things.

Reply:
If it is something to be presented to the admin, and we only need this info for some time, then perhaps using `group0_history` is indeed a valid approach.

Reply:
How long does the `group0_history` entries stick around. Are there conditions under which entries can be cleaned up very quickly after being written?

Reply:
Entry X is cleared if it is older than 1 week when new entry is added (i.e. group 0 command is applied)
this is controlled by `raft_group0_client::_history_gc_duration`
```cpp
    gc_clock::duration _history_gc_duration = gc_clock::duration{std::chrono::duration_cast<gc_clock::duration>(std::chrono::weeks{1})};
```

```cpp
group0_command raft_group0_client::prepare_command(Command change, group0_guard& guard, std::string_view description) {
    group0_command group0_cmd {
        .change{std::move(change)},
        .history_append{db::system_keyspace::make_group0_history_state_id_mutation(
            guard.new_group0_state_id(), _history_gc_duration, description)},
```

```cpp
mutation system_keyspace::make_group0_history_state_id_mutation(
        utils::UUID state_id, std::optional<gc_clock::duration> gc_older_than, std::string_view description) {
    auto s = group0_history();
    mutation m(s, partition_key::from_singular(*s, GROUP0_HISTORY_KEY));
    auto& row = m.partition().clustered_row(*s, clustering_key::from_singular(*s, state_id));
    auto ts = utils::UUID_gen::micros_timestamp(state_id);
    row.apply(row_marker(ts));
    if (!description.empty()) {
        auto cdef = s->get_column_definition("description");
        assert(cdef);
        row.cells().apply(*cdef, atomic_cell::make_live(*cdef->type, ts, cdef->type->decompose(description)));
    }
    if (gc_older_than) {
        using namespace std::chrono;
        assert(*gc_older_than >= gc_clock::duration{0});

        auto ts_micros = microseconds{ts};
        auto gc_older_than_micros = duration_cast<microseconds>(*gc_older_than);
        assert(gc_older_than_micros < ts_micros);

        auto tomb_upper_bound = utils::UUID_gen::min_time_UUID(ts_micros - gc_older_than_micros);
        // We want to delete all entries with IDs smaller than `tomb_upper_bound`
        // but the deleted range is of the form (x, +inf) since the schema is reversed.
        auto range = query::clustering_range::make_starting_with({
                clustering_key_prefix::from_single_value(*s, timeuuid_type->decompose(tomb_upper_bound)), false});
        auto bv = bound_view::from_range(range);

        m.partition().apply_delete(*s, range_tombstone{bv.first, bv.second, tombstone{ts, gc_clock::now()}});
    }
    return m;
}
```

Reply:
> Are there conditions under which entries can be cleaned up very quickly after being written?

I guess manual recovery procedure which requires truncating group0_history among other tables but that's only for disaster scenarios (group 0 majority loss)
https://opensource.docs.scylladb.com/master/troubleshooting/handling-node-failures.html#manual-recovery-procedure

Reply:
Type is supposed to be human readable, so I think `group0_history` would be enough. If users request info about the record that is truncated, they may get "type details are lost" type I think

Reply:
A TTL of one week should be more then enough for the purposes of task-manager. @Deexie

Reply:
There is a problem with getting operation type from `group0_history` - joining uses two different ids in `group0_history` and `topology_requests`. The latter (and also node ops tasks) uses request_id generated in `storage_service::join_token_ring`, the former uses state id generated using `generate_group0_state_id` (in `storage_service::join_node_request_handler`).

I don't think we can make `state_id = request_id` as it needs to be strictly greater than previous `state_id`. We could theoretically make `request_id = state_id`, but we would need to change `join_node_request` rpc return type, which again requires version synchronization. I don't think there is another way to translate the two.

So, maybe it's better to stay with a new column. What do you think? @denesb @kbr-scylla @gleb-cloudius

Reply:
> There is a problem with getting operation type from `group0_history` - joining uses two different ids in `group0_history` and `topology_requests`. The latter (and also node ops tasks) uses request_id generated in `storage_service::join_token_ring`, the former uses state id generated using `generate_group0_state_id` (in `storage_service::join_node_request_handler`).
> 
> I don't think we can make `state_id = request_id` as it needs to be strictly greater than previous `state_id`. We could theoretically make `request_id = state_id`, but we would need to change `join_node_request` rpc return type, which again requires version synchronization. I don't think there is another way to translate the two.
> 
> So, maybe it's better to stay with a new column. What do you think? @denesb @kbr-scylla @gleb-cloudius

If version synchronization is needed anyway adding a column is better.